cmd/gpu_plugin/rm/gpu_plugin_resource_manager.go

```diff
@@ -120,45 +139,274 @@ func NewResourceManager(skipID, fullResourceName string) (ResourceManager, error
+		klog.Error("couldn't find allocation info from assignments:", getPodKey(pod))
+
+		return nil, &dpapi.UseDefaultMethodError{}
+	}
+
+	containerIndex := podCandidate.allocatedContainers
+
+	affinityMask := assignment.containers[containerIndex].tileEnv
+	getPrefDevices := assignment.containers[containerIndex].deviceIds
+
+	rm.cleanupMutex.Unlock()
+
+	devIds := request.ContainerRequests[0].DevicesIDs
+
+	// Check if all the preferred devices were also used
+	if len(devIds) != len(getPrefDevices) {
+		klog.Warningf("Allocate called with odd number of device IDs: %d vs %d", len(devIds), len(getPrefDevices))
+	}
+
+	for _, devID := range devIds {
+		if _, found := getPrefDevices[devID]; !found {
+			klog.Warningf("Not preferred device used in Allocate: %s (%v)", devID, getPrefDevices)
+		}
+	}
+
+	klog.V(4).Info("Allocate affinity mask: ", affinityMask)
+	klog.V(4).Info("Allocate device ids: ", devIds)
+
+	return rm.createAllocateResponse(devIds, affinityMask)
+}
+
+func (rm *resourceManager) GetPreferredFractionalAllocation(request *pluginapi.PreferredAllocationRequest) (
+	*pluginapi.PreferredAllocationResponse, error) {
+	response := pluginapi.PreferredAllocationResponse{}
+
+	if !isPreferredAllocationRequestOk(request, rm.skipID) {
+		// it is better to leave allocated gpu devices as is and return
+		return &response, nil
+	}
+
+	klog.V(4).Info("GetPreferredAllocation request: ", request)
+
+	podCandidate, err := rm.findAllocationPodCandidate()
+	if errors.Is(err, &retryErr{}) {
+		klog.Warning("retrying POD resolving after sleeping")
+		time.Sleep(rm.retryTimeout)
+
+		podCandidate, err = rm.findAllocationPodCandidate()
+	}
+
+	if err != nil {
+		if !errors.Is(err, &zeroPendingErr{}) {
+			klog.Error("allocation candidate not found, perhaps the GPU scheduler extender is not called, err:", err)
+		}
+
+		// Return empty response as returning an error causes
+		// the pod to be labeled as UnexpectedAdmissionError
+		return &response, nil
+	}
+
+	pod := podCandidate.pod
+	allocatedContainers := podCandidate.allocatedContainers
+	cards := containerCards(pod, allocatedContainers)
+	affinityMask := containerTileAffinityMask(pod, allocatedContainers)
+	podKey := getPodKey(pod)
+
+	creq := request.ContainerRequests[0]
+
+	klog.V(4).Info("Get preferred fractional allocation: ",
+		podKey, creq.AllocationSize, creq.MustIncludeDeviceIDs, creq.AvailableDeviceIDs)
+
+	deviceIds := selectDeviceIDsForContainer(
+		int(creq.AllocationSize), cards, creq.AvailableDeviceIDs, creq.MustIncludeDeviceIDs)
+
+	// Map container assignment details per pod name
+
+	rm.cleanupMutex.Lock()
+
+	assignments, found := rm.assignments[podKey]
+
+	if !found {
+		assignments.containers = make([]ContainerAssignments, podCandidate.allocationTargetNum)
+	}
+
+	assignments.containers[allocatedContainers].tileEnv = affinityMask
+	// Store device ids so we can double check the ones in Allocate
+	assignments.containers[allocatedContainers].deviceIds = make(map[string]bool)
```

Comment:
Is `assignments.containers[allocatedContainers]` special? Shouldn't we assign `tileEnv` and `deviceIds` something for the other containers?

Reply:
Yeah, I'll rename it as the current one is confusing.